Regularized Gaussian extension: TV

jeverink opened the pull request:
Adds Total Variation regularization as option to the regularized Gaussian distribution.

PR features:

- [x] Total Variation ("tv") option with RegularizedGaussian and its variants. (1D and anisotropic 2D, depending on passed geometry)
- [x] Conditioning on the strength parameter for RegularizedGaussian (for future use with Gibbs)

Regarding tests: I propose adding a small regression test of a full hierarchical model with a TV regularized model as part of the next PR, as many smaller things might change rapidly in future smalles PRs on the Regularized Gaussian Distribution.

Example results:

`
n = 128
A, y_data, info = Deconvolution1D(dim=n, phantom='square').get_components()

x  = RegularizedGaussian(0.5*np.ones(n), 0.1, regularization = "tv", strength = 10)
y  = Gaussian(A@x, 0.001)

joint = JointDistribution(x, y)
posterior = joint(y=y_data)

sampler = RegularizedLinearRTO(posterior,
                                maxit=50,
                                penalty_parameter = 10) 
sampler.sample(100, 10)

samples = sampler.get_samples()
`

![image](https://github.com/user-attachments/assets/0ff69ba4-b123-4e09-940c-3afa2cf43d6d)

`
M = 100
A, y_data, info = Deconvolution2D(dim=M, phantom="cookie").get_components()

x  = RegularizedGMRF(0.0*np.ones(M**2), prec = lambda d:d, regularization = "tv", strength = lambda d:d, geometry = A.domain_geometry)
x = x(d = 10)
y  = Gaussian(A@x, 0.001)

joint = JointDistribution(x, y)
posterior = joint(y=y_data)

sampler = RegularizedLinearRTO(posterior,
                                maxit=25,
                                penalty_parameter = 10) 
sampler.sample(100, 10)

samples = sampler.get_samples()
`

![image](https://github.com/user-attachments/assets/5d686831-175f-4eb4-8c7f-60ef606ebdf7)

nabriis commented on cuqi/experimental/mcmc/_rto.py:
```diff
@@ -197,10 +204,13 @@ def __init__(self, target=None, initial_point=None, maxit=100, stepsize="automat
     def _initialize(self):
         super()._initialize()
-        self._stepsize = self._choose_stepsize()
+        if callable(self.proximal):
```
This logic is possibly difficult to understand in the future. I suggest you add an attribute like "_inner_solver" that can be either ADMM or FISTA (can just be a string or likewise). Then use checks like that to see which inner solver is used. Define the inner solver when the target is set.

Same holds for RegularizedGausssian

jeverink replied:
I have now added the _inner_solver attribute.

I do not understand what you mean with 
> Same holds for RegularizedGausssian

nabriis replied:
Thanks. Looks good. In both classes.

nabriis commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -57,10 +64,10 @@ class RegularizedGaussian(Distribution):
-        Preset regularization. Can be set to "l1". Required for use in Gibbs in future update.
-        For "l1", the following additional parameters can be passed:
+        Preset regularization. Can be set to "l1" or 'tv'. Required for use in Gibbs in future update.
+        For "l1" or "tv", the following additional parameters can be passed:
```
Is the zero boundary conditions described anywhere for TV? else I suggest we add it in the documentation.

jeverink replied:
Good catch. I have now changed the boundary condition. Weirdly the boundary condition had to be set to "neumann" for there to be no boundary conditions, instead of "none".

nabriis commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -113,15 +128,50 @@ def _parse_regularization_input_arguments(self, proximal, projector, constraint,
+            self._strength = optional_regularization_parameters["strength"]
+            if isinstance(self.geometry, (Continuous1D)):
+                self._transformation = FirstOrderFiniteDifference(self.geometry.par_dim, bc_type='zero')
+            elif isinstance(self.geometry, (Continuous2D, Image2D)):
```
If Image2D geometry has attribute "visual_only" as true I am not sure we want to actually use a 2D TV prior. Perhaps raise an error in this case for now.

jeverink replied:
I have now added this condition

nabriis commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -206,16 +256,18 @@ def sqrtcov(self):
         self.gaussian.sqrtcov = value     
     
-    def get_conditioning_variables(self):
-        return self.gaussian.get_conditioning_variables()
```
Why is this removed?

jeverink replied:
The first thing that `get_conditioning_variables` does is ask for `get_mutable_variables`. So as long as `get_mutable_variables` and the attributes are correct, this wrapping should not be needed anymore.

nabriis replied:
You are right!

nabriis commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -206,16 +256,18 @@ def sqrtcov(self):
-    
     def get_mutable_variables(self):
-        return self.gaussian.get_mutable_variables()
+        add = []
```
Cool to have strength as a mutable variable! I suggest this a bit refactoring
```python
def get_mutable_variables(self):
    mutable_vars = self.gaussian.get_mutable_variables()
    
    if self.preset in self.regularization_options():
        mutable_vars += ["strength"]
    
    return mutable_vars
```

jeverink replied:
Thanks for the suggestion.

nabriis commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -206,16 +256,18 @@ def sqrtcov(self):
     # which means we convert Distribution to Likelihood or EvaluatedDensity.
     def _condition(self, *args, **kwargs):
-
+        if self.preset in self.regularization_options():
+            return super()._condition(*args, **kwargs)
```
There seem to be some potential issues here. Namely, what if both the mean and strength were conditioning variables? Then we would only condition on the strength now. E.g.
```python
x = RegularizedGaussian(lambda x: x, 1, strength=lambda s: s, ...) 
x(strength=1, mean=2)
```
I suggest moving this check after creating "new_density" at the end, then doing the super()._condition(*args, **kwargs) on new_density.

It could also be worthwhile to add a test on this new conditioning feature.

jeverink replied:
The precise workings of the conditioning is still a little bit of a mystery to me, but the issue you suggest, where there are anonymous functions for both Gaussian parameters and regularization parameters, seems to be working just fine. Also, in the full version of the code that I have, including conjugacy, this seems to work as well. 

![image](https://github.com/user-attachments/assets/12abc9b8-133e-47d3-aea1-0247f7459ba0)



I do get that this is not the best solution, however your suggestion,
if I interpret it as removing the initial check and replacing the final line with
`return new_density._condition(**kwargs)` crashes the kernel.

I have now added two tests, one with constraints and one with tv, to test the mutable and conditioning variables and verify that the conditioning does work as expected.

nabriis replied:
Thank you. I look into it a bit more. It seems to be working as expected though

nabriis reviewed (changes requested):
Hi @jeverink. Really nice and clean PR. I only had some minor comments on docstrings and then one main area for comments on the code for conditioning which was now updated. Let me know if you have any questions

jeverink commented on cuqi/experimental/mcmc/_rto.py:
```diff
@@ -210,8 +220,8 @@ def validate_target(self):
-        if not callable(self.proximal):
-            raise TypeError("Proximal needs to be callable")
+        if not callable(self.proximal) and not isinstance(self.proximal, list):
+            raise TypeError("Proximal needs to be callable or a list")
```
I have updated the error message

jeverink commented:
Thanks for the feedback @nabriis ,

The conditioning of the regularized Gaussian distributions is still a little bit of a mystery to me, but it all does seem to work. I hope you have a better understanding of it to know how to properly improve it.

nabriis commented on cuqi/experimental/mcmc/_rto.py:
```diff
@@ -169,12 +169,19 @@ class RegularizedLinearRTO(LinearRTO):
         Initial point for the sampler. *Optional*.
```
Great thanks!

amal-ghamdi commented on tests/test_implicit_priors.py:
```diff
@@ -104,3 +104,37 @@ def test_RegularizedUnboundedUniform_is_RegularizedGaussian():
+    assert np.allclose(x.mean, [1, 1, 1, 1])
+    assert np.allclose(x.prec, 10)
+
+def test_RegularizedGaussian_conditioning_strength():
```
Nice!

amal-ghamdi reviewed (changes requested):
Thank you @jeverink for the great work here and expanding the features of the regularized linear RTO! I have some comments here and there for you to address as you see fit. 

One additional aspect I wanted to mention, if there is a way to simplify setting constriants/projection/ etc. Now we have: "proximal, projector, constraint, regularization" where proximal can be a list. Can everything be unified in one list for example? If this makes sense to you but is a big change, I suggest adding your thoughts on it in an issue.

jeverink commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -39,10 +41,15 @@ class RegularizedGaussian(Distribution):
+        If list of tuples (callable proximal operator of f_i, linear operator L_i):
+            Each callable proximal operator of f_i accepts two arguments (x, p) and should return the minimizer of p/2||x-z||^2 + f(x) over z for some f.
+            The corresponding regularization takes the form
+                sum_i f_i(L_i x),
```
This is incorrect.

jeverink commented on cuqi/experimental/mcmc/_rto.py:
```diff
@@ -210,8 +224,9 @@ def validate_target(self):
-        if not callable(self.proximal):
-            raise TypeError("Proximal needs to be callable")
+        if not callable(self.proximal) and not isinstance(self.proximal, list):
+            raise TypeError("Proximal in the regularized Gaussian prior needs to be callable or a list")
```
When it is a list, it is specified in the exact same way as for ADMM, as this is what is passed on to ADMM.

But I agree that this error message is superfluous, given that we check whether the prior is a RegularizedGaussian or RegularizedGMRF directly above, which constructs and/or verifies these precise conditions.

jeverink commented on cuqi/experimental/mcmc/_rto.py:
```diff
@@ -159,6 +159,10 @@ class RegularizedLinearRTO(LinearRTO):
     Samples posterior related to the inverse problem with Gaussian likelihood and implicit Gaussian prior, and where the forward model is Linear.
+    The sampler works by repeatedly solving regularized linear least squares problems.
+    The solver for these optimization problems is chosen based on how the regularized is provided in the implicit Gaussian prior.
+    See :class:`~cuqi.implicitprior.RegularizedGaussian` for details.
```
I have now replace it with a comment on the conditions under which solver is used.

jeverink commented on cuqi/implicitprior/_regularizedGaussian.py:
```diff
@@ -89,10 +97,17 @@ def _parse_regularization_input_arguments(self, proximal, projector, constraint,
+                    if op.shape[1] != self.geometry.par_dim:
+                        raise ValueError("Incorrect shape of linear operator.")
+            else:
+                raise ValueError("Proximal needs to be callable or a list. See documentation.")
```
I have now created a setter for the proximal with these checks, and made sure it works with the presets.

jeverink commented:
Thanks for the feedback @amal-ghamdi ,
regarding the possible unification of the penalty terms,
I think that is a great idea to streamline the features, I have added the issue #584.

Also to @nabriis ,
I have now added a general issue #585 with a bit of a roadmap of the upcoming/planned features that will hopefully follow from this PR.

amal-ghamdi reviewed (approved):
Thank you @jeverink for the updates. I like how you added more details to the documentation to clarify more concepts to the user, especially in what concerns constraints/regularizers/proximals. Also thanks for creating #584 .

nabriis commented:
@jeverink 🎉
